core/src/main/java/com/google/errorprone/bugpatterns/inject/dagger/UsageCheckers.java

```diff
@@ -0,0 +1,126 @@
+package com.google.errorprone.bugpatterns.inject.dagger;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.BugPattern;
+
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.util.TreePathScanner;
+import com.sun.source.tree.IdentifierTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.HashMap;
+
+import static com.google.errorprone.BugPattern.Category.DAGGER;
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+
+/** @authors Murali Krishna Ramanathan, Manu Sridharan */
+/** Class that analyzes the use of variables to determine unused variables. */
+@AutoService(BugChecker.class)
+@BugPattern(
+  name = "UsageCheckers",
+  altNames = {"UnusedProviderParam", "UnusedMemberInjection"},
+  summary = "Unused field with @Inject annotation.",
+  category = DAGGER,
+  severity = WARNING
+)
+public class UsageCheckers extends BugChecker implements BugChecker.ClassTreeMatcher {
+
+  private static final String unusedInjectCheckerName = "UnusedMemberInjection";
+  private static final String unusedInjectCheckerMessage = "Unused field with @Inject annotation.";
+  private static final String unusedParamCheckerName = "UnusedProviderParam";
+  private static final String unusedParamCheckerMessage =
+      "Unused parameter in method with @Provides annotation.";
+
+  private boolean symbolHasSuppressInitalizationWarningsAnnotation(
+      Symbol symbol, String checkerName) {
+    SuppressWarnings annotation = symbol.getAnnotation(SuppressWarnings.class);
+    if (annotation != null) {
+      for (String s : annotation.value()) {
+        if (s.equals(checkerName)) return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public Description matchClass(ClassTree tree, VisitorState state) {
+    UsageCheckers.CallScanner callScanner = new UsageCheckers.CallScanner(state);
+    callScanner.scan(state.getPath(), null);
```

Comment:
I think we should invert this. We should be cautious that this will run on every Java class, so creating maps of every variable usage could be very wasteful.

Instead, I think we should have individual matchers for (a) fields and methods with `@Inject` and (b) methods with `@Provides` (and we should also add `@dagger.producers.Produces` too) and only when we find those should we then check to see if they're used. Then we probably don't have to do any state keeping of what is used or not - we can scan looking for a single symbol and once we find it, stop scanning. If it's never found, then we declare it unused.